### PR TITLE
Update Clifford Circuit Learning MitEx

### PR DESCRIPTION
**Pull request by sjdilkes**
This pull request makes several changes to the `TaskGraph` returned by `gen_CCL_MitEx`. This is to account for some state circuits producing noiseless and noisy expectation values that ultimately perturbed the model and correction significantly.

1. Add Barrier gates around every Rz operation. In some cases Clifford angles could be sampled that could be easily optimised into the rest of the circuit, reducing Circuit depth massively and leaving very accurate noisy expectation values where other Circuits had worse ones. Barrier gates stop default optimisation (even at level 0) from modifying circuit structure.
2. Add edge cases for how state circuits are constructed for gates with few Rz ops. Also improve logic for automatically determining sampled pairs and number of assigned "Clifford" and "Non-Clifford" Rz gates in state circuit generation.
3. Add tolerance to which expectation values are ultimately added to a model. This functions like an additional `LikelihoodFunction`, but skimming individual `QubitPauliString` in `QubitPauliOperator` instead. If an exact expectation value is 0 (or within tolerance) its inclusion to the model can cause the model to be ineffective.

### CODE REVIEW

File: qermit/clifford_noise_characterisation/ccl.py [hunk 1]
```diff
@@ -151,26 +151,50 @@ def gen_state_circuits(
...
+            # all circuits accepted and run, some results later discarded if not accepted by Metropolis-Hastings rule
+            state_circuits.append(new_circuit)
+        return state_circuits
+
+    n_non_cliffords = min(n_non_cliffords, len(rz_ops) - 1)
+    n_cliffords = len(rz_ops) - n_non_cliffords
+
+    # if n_cliffords == 0:
+    #     n_non_cliffords = n_non_cliffords - 1
+    #     n_cliffords = 1
```
Comment (daniel-mills-cqc) [inline qermit/clifford_noise_characterisation/ccl.py:186]: Can this be removed?

File: qermit/clifford_noise_characterisation/ccl.py [hunk 2]
```diff
@@ -461,7 +491,7 @@ def gen_CDR_MitEx(
             MitEx(
                 device_backend,
                 _label="ExperimentMitex",
-                mitres=gen_compiled_MitRes(device_backend, 2),
+                mitres=gen_compiled_MitRes(device_backend, 0),
```
Comment (daniel-mills-cqc) [inline qermit/clifford_noise_characterisation/ccl.py:494]: Is it necessary to switch to optimisation level 0 given that the barriers have been added?

**Review by daniel-mills-cqc — CHANGES_REQUESTED** [review]
Looking good, just a couple of small comments from me.

File: qermit/clifford_noise_characterisation/ccl.py (hunk 1 above)
Comment (sjdilkes) [reply qermit/clifford_noise_characterisation/ccl.py:186]: yeahp! thanks

File: qermit/clifford_noise_characterisation/ccl.py (hunk 2 above)
Comment (sjdilkes) [reply qermit/clifford_noise_characterisation/ccl.py:494]: Unfortunately this still seems to be a requirement for consistent results

**Review by daniel-mills-cqc — APPROVED** [review]

**Pull request merged**